server/src/main/java/org/opentosca/toscana/plugins/cloudformation/visitor/CloudFormationNodeVisitor.java

```diff
@@ -45,85 +54,118 @@
+     */
+    protected static Compute getCompute(WebApplication webApplication) {
+        WebServer webServer = webApplication.getHost().getNode().orElseThrow(
+            () -> new IllegalStateException("WebApplication is missing WebServer")
+        );
+        return getCompute(webServer);
+    }
+
+    /**
+     Get the Compute node this mysqlDatabase is ultimately hosted on
+
+     @param database the mysqlDatabase to find the host for
+     @return the underlying Compute node
+     */
+    protected static Compute getCompute(Database database) {
+        Dbms dbms = database.getHost().getNode().orElseThrow(
+            () -> new IllegalStateException("MysqlDatabase is missing Dbms")
+        );
+        return dbms.getHost().getNode().orElseThrow(
+            () -> new IllegalStateException("Dbms is missing Compute")
+        );
+    }
+
+    /**
+     Get the Compute node this webServer is ultimately hosted on
+
+     @param webServer the mysqlDatabase to find the host for
+     @return the underlying Compute node
+     */
+    protected static Compute getCompute(WebServer webServer) {
+        return webServer.getHost().getNode().orElseThrow(
+            () -> new IllegalStateException("WebServer is missing Compute")
+        );
+    }
+
     @Override
     public void visit(Compute node) {
         try {
-            logger.debug("Visit Compute node " + node.getEntityName() + ".");
-            String nodeName = toAlphanumerical(node.getEntityName());
-            //default security group the EC2 Instance opens for port 80 and 22 to the whole internet
-            Object cidrIp = "0.0.0.0/0";
-            SecurityGroup webServerSecurityGroup = cfnModule.resource(SecurityGroup.class,
-                nodeName + SECURITY_GROUP)
-                .groupDescription("Enable ports 80 and 22")
-                .ingress(ingress -> ingress.cidrIp(cidrIp), "tcp", 80, 22);
-
-            // check what image id should be taken
-            CapabilityMapper capabilityMapper = createCapabilityMapper();
-
-            OsCapability computeOs = node.getOs();
-            String imageId = capabilityMapper.mapOsCapabilityToImageId(computeOs);
-            ComputeCapability computeCompute = node.getHost();
-            String instanceType = capabilityMapper.mapComputeCapabilityToInstanceType(computeCompute,
-                CapabilityMapper.EC2_DISTINCTION);
-            //create CFN init and store it
-            CFNInit init = new CFNInit(CONFIG_SETS);
-            cfnModule.putCFNInit(nodeName, init);
-            //takes default 8GB storage but volume
-            cfnModule.resource(Instance.class, nodeName)
-                .keyName(cfnModule.getKeyNameVar())
-                .securityGroupIds(webServerSecurityGroup)
-                .imageId(imageId)
-                .instanceType(instanceType);
+            logger.info("Visit Compute node '{}'.", node.getEntityName());
+            if (cfnModule.checkComputeToEc2(node)) {
```

Comment:
Maybe I'm misunderstanding this, but doesn't this mean that a Compute node only gets transformed if an Apache is hosted on it? Since `addComputeToEc2` only gets called in the `visit()` method of Apache in `PrepareModelNodeVisitor`.

Reply:
Yeah this is wrong, I changed it